samples/StaticFilesAuth/Startup.cs

```diff
@@ -0,0 +1,192 @@
+                var usersPath = Path.Combine(basePath, "Users");
+
+                options.AddPolicy("files", builder =>
+                {
+                    builder.RequireAuthenticatedUser().RequireAssertion(context =>
+                    {
+                        // Users are only authorized to access the base directory, the Users directory, and their own directory under Users.
+                        var userName = context.User.Identity.Name;
+                        userName = userName?.Split('@').FirstOrDefault();
+                        if (userName == null)
+                        {
+                            return false;
+                        }
+                        var userPath = Path.Combine(usersPath, userName);
+                        if (context.Resource is IFileInfo file)
+                        {
+                            var path = Path.GetDirectoryName(file.PhysicalPath);
+                            return string.Equals(path, basePath, StringComparison.OrdinalIgnoreCase)
+                                || string.Equals(path, usersPath, StringComparison.OrdinalIgnoreCase)
+                                || string.Equals(path, userPath, StringComparison.OrdinalIgnoreCase)
+                                || path.StartsWith(userPath + Path.DirectorySeparatorChar, StringComparison.OrdinalIgnoreCase);
+                        }
+                        else if (context.Resource is IDirectoryContents dir)
+                        {
+                            // https://github.com/aspnet/Home/issues/3073
+                            // This won't work right if the directory is empty
+                            var path = Path.GetDirectoryName(dir.First().PhysicalPath);
+                            return string.Equals(path, basePath, StringComparison.OrdinalIgnoreCase)
+                                || string.Equals(path, usersPath, StringComparison.OrdinalIgnoreCase)
+                                || string.Equals(path, userPath, StringComparison.OrdinalIgnoreCase)
+                                || path.StartsWith(userPath + Path.DirectorySeparatorChar, StringComparison.OrdinalIgnoreCase);
+                        }
+
+                        throw new NotImplementedException($"Unknown resource type '{context.Resource.GetType()}'");
+                    });
+                });
+            });
+
+            services.AddMvc().SetCompatibilityVersion(CompatibilityVersion.Version_2_1);
+        }
+
+        // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
+        public void Configure(IApplicationBuilder app, IHostingEnvironment env, IAuthorizationService authorizationService)
+        {
+            if (env.IsDevelopment())
+            {
+                app.UseDeveloperExceptionPage();
+            }
+            else
+            {
+                app.UseExceptionHandler("/Home/Error");
+            }
+
+            // Public
```

Comment:
I feel it would be good to add more description than this. I suppose StaticFiles always (?) looks only in `wwwroot` folder and all files in it are public.